<a href="https://colab.research.google.com/github/hzhoujoy/ST554_HW7/blob/main/JoyZhou_HW7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Home Work 7**   
Author: Huiping Zhou  
Date: 3/28/2026  

# Instruction
The purpose of this homework is to practice fitting MLR and logistic regression models (including penalized or regularized models).
We will use a ` Wine Quality` dataset from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/186/wine+quality).
- fixed_acidity
- volatile_acidity
- citric_acid
- residual_sugar
- chlorides
- free_sulfur_dioxide
- total_sulfur_dioxide
- density
- pH
- sulphates
- alcohol
- quality: score between 0 to 10
- color: red or white   
we will use `alcohol` as a target variable for fitting multiple linear regression model while using `type of wine` as the response variable for fitting logistic regression model.

## Read in and Combine Data  
We start by downloading the `winequality-red.csv` and `winequality-white.csv` datasets from the [UCI machine
learning repository site](https://archive.ics.uci.edu/dataset/186/wine+quality), saving them locally, and then uploaingd them into the notebook.

Next, we use the `pd.read_cvs()` function with relative file paths to read in the two datasets.


In [1]:
import pandas as pd
import numpy as np
# read in winequality-red dataset
red_data = pd.read_csv("winequality-red.csv", sep=";")
#Display summary information
red_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


According to the info() output, `red_data` includes 1,599 rows and 12 variables.

In [2]:
#check the first 5 rows of dataset
red_data.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5


In [3]:
#read in dataset
white_data = pd.read_csv("winequality-white.csv", sep=";")

#provides a summary of the dataset
white_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


According to the info() output, `white_data` includes 4,898 rows and 12 columns.

In [4]:
#check the first 5 rows of dataset
white_data.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
0,7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.0010,3.00,0.45,8.8,6
1,6.3,0.30,0.34,1.6,0.049,14.0,132.0,0.9940,3.30,0.49,9.5,6
2,8.1,0.28,0.40,6.9,0.050,30.0,97.0,0.9951,3.26,0.44,10.1,6
3,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6
4,7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.40,9.9,6


Then, we combine these two datasets and create a new variable that represents the type of wine (red or white).

In [5]:
# Combine along columns
wine_df = pd.concat([red_data, white_data], ignore_index=True)
wine_df.info() #check summary information of combined dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 609.2 KB


According to the info() output, combined `wine_df` includes 6497 rows and 12 columns, indicating that the red and white wine datasets were merged properly.

We create a new variable, `type_of_wine`, to indicate whether each observation corresponds to red or white wine type in the combined dataset.

In [6]:
#First part of the combined dataset (rows originally from red_data): label as "red"
wine_df.loc[:len(red_data)-1, "type_of_wine"] = "red"

# Remaining rows (originally from white_data): label as "white"
wine_df.loc[len(red_data):, "type_of_wine"] = "white"

In [7]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         6497 non-null   float64
 1   volatile acidity      6497 non-null   float64
 2   citric acid           6497 non-null   float64
 3   residual sugar        6497 non-null   float64
 4   chlorides             6497 non-null   float64
 5   free sulfur dioxide   6497 non-null   float64
 6   total sulfur dioxide  6497 non-null   float64
 7   density               6497 non-null   float64
 8   pH                    6497 non-null   float64
 9   sulphates             6497 non-null   float64
 10  alcohol               6497 non-null   float64
 11  quality               6497 non-null   int64  
 12  type_of_wine          6497 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


We update the column names in the `wine_df` DataFrame by replacing spaces with underscores.

In [8]:
wine_df.columns = [col.replace(" ", "_") for col in wine_df.columns]
wine_df.head()

,fixed_acidity,volatile_acidity,citric_acid,residual_sugar,chlorides,free_sulfur_dioxide,total_sulfur_dioxide,density,pH,sulphates,alcohol,quality,type_of_wine
0,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red
1,7.8,0.88,0.00,2.6,0.098,25.0,67.0,0.9968,3.20,0.68,9.8,5,red
2,7.8,0.76,0.04,2.3,0.092,15.0,54.0,0.9970,3.26,0.65,9.8,5,red
3,11.2,0.28,0.56,1.9,0.075,17.0,60.0,0.9980,3.16,0.58,9.8,6,red
4,7.4,0.70,0.00,1.9,0.076,11.0,34.0,0.9978,3.51,0.56,9.4,5,red


## Regression Task (`alcohol` as response)   
We first fit the regression models using `alcohol` as the response variable.

## Split the data

In [11]:
import numpy as np
from sklearn.model_selection import train_test_split

#Split the dataset into predictors (X) and response (y)
X_train, X_test, y_train, y_test = train_test_split(
    wine_df.drop(['alcohol', 'type_of_wine'], axis=1), #prodictors(X)
    wine_df['alcohol'],                                #response
    test_size=0.3,                                     #70/30 split
    random_state=42,
    stratify=wine_df['type_of_wine']                   #maintains wine type proportions in each split
)